### PR TITLE
seccomp-util: pass negative fds as is to fsync() and friends

### LINKED ISSUE #34478: systemd-nspawn `--suppress-sync=true` also suppresses fd validation in `fsync()`
### systemd version the issue has been seen with

256.5

### Used distribution

Gentoo Linux

### Linux kernel version used

6.10.10-gentoo-dist

### CPU architectures issue was seen on

x86_64

### Component

systemd-nspawn

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

`fsync()` (and related calls) returns 0 when passed invalid fd and `--suppress-sync=true` is used.

This is causing CPython's test suite to fail.

### Steps to reproduce the problem

1. Enter a container with `systemd-nspawn --suppress-sync=true`.
2. Compile and run the following program:
    ```c
    #include <stdio.h>
    #include <unistd.h>
    
    int main() {
            printf("fdatasync() = %d\n", fdatasync(-1));
            perror("errno");
            return 0;
    }
    ```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
fdatasync = 0
errno: Success

(while without --suppress-sync=true, it gives:
fdatasync = -1
errno: Bad file descriptor
)
```

### PR DESCRIPTION
**Pull request by yuwata**
Closes #34478.

### CODE REVIEW

**Comment by YHNdnzj** [comment]
This is now included in #34492.

**Comment by yuwata** [comment]
Uh, oh, you are working on the same issue...

**Comment by yuwata** [comment]
Now this works fine. And it seems you are in the same pitfall that I spend few hours...

**Comment by yuwata** [comment]
@YHNdnzj Dropping the replaced tag. As now this works.

File: src/test/test-seccomp.c [hunk 1]
```diff
@@ -1229,4 +1230,56 @@ TEST(restrict_suid_sgid) {
...
+                log_notice("Seccomp not available, skipping %s", __func__);
+                return;
+        }
+        if (!have_seccomp_privs()) {
+                log_notice("Not privileged, skipping %s", __func__);
+                return;
+        }
+
+        pid = fork();
+        assert_se(pid >= 0);
```
Comment (YHNdnzj) [inline src/test/test-seccomp.c]: ASSERT_OK_ERRNO

File: src/shared/seccomp-util.c [hunk 2]
```diff
@@ -2483,11 +2484,21 @@ int seccomp_suppress_sync(void) {
...
-                                        SCMP_ACT_ERRNO(0), /* success → we want this to be a NOP after all */
-                                        id,
-                                        0);
+                        if (STR_IN_SET(c, "fdatasync", "fsync", "sync_file_range", "sync_file_range2", "syncfs"))
+                                r = seccomp_rule_add_exact(
+                                                seccomp,
+                                                SCMP_ACT_ERRNO(0), /* success → we want this to be a NOP after all */
+                                                id,
+                                                1,
+                                                SCMP_A0(SCMP_CMP_LE, INT_MAX)); /* The rule handles arguments in unsigned. Hence, this
```
Comment (YHNdnzj) [inline src/shared/seccomp-util.c:2495]: Ugh, I always forget that kernel works with unsigned fds

File: src/shared/seccomp-util.c [hunk 3]
```diff
@@ -2483,11 +2484,21 @@ int seccomp_suppress_sync(void) {
                                 continue;
                         }
 
-                        r = seccomp_rule_add_exact(
-                                        seccomp,
-                                        SCMP_ACT_ERRNO(0), /* success → we want this to be a NOP after all */
-                                        id,
-                                        0);
+                        if (STR_IN_SET(c, "fdatasync", "fsync", "sync_file_range", "sync_file_range2", "syncfs"))
```
Comment (YHNdnzj) [inline src/shared/seccomp-util.c:2489]: Please also update the comment in the beginning of the function: https://github.com/systemd/systemd/pull/34492/commits/73d390be3cb2360d7963aafde684d45cf67d7a86#diff-60141857536525beb87708eaf458438b24ce034aed48172f0db7ee1cd1acc648R2476

File: src/shared/seccomp-util.c [hunk 4]
```diff
@@ -2483,11 +2484,21 @@ int seccomp_suppress_sync(void) {
...
+                                r = seccomp_rule_add_exact(
+                                                seccomp,
+                                                SCMP_ACT_ERRNO(0), /* success → we want this to be a NOP after all */
+                                                id,
+                                                1,
+                                                SCMP_A0(SCMP_CMP_LE, INT_MAX)); /* The rule handles arguments in unsigned. Hence, this
+                                                                                 * means non-negative fd matches the rule, and the
+                                                                                 * negative fd passed to the syscall (then it fails EBADF). */
```
Comment (YHNdnzj) [inline src/shared/seccomp-util.c]: fails with

**Review by YHNdnzj — APPROVED** [review]

**Pull request merged**